### ISSUE TITLE
docker module fails to restart container

### ISSUE THREAD

**Issue by andrenarchy** [ansible/ansible-modules-core#2824]
##### ISSUE TYPE

Bug Report
##### COMPONENT NAME

_docker module
##### ANSIBLE VERSION

2.0.0.2
##### SUMMARY

Ansible 2.0.0.2 fails to restart a docker container with the error message

```
RUNNING HANDLER [docker-registry : Restart docker registry] ********************
fatal: [dr-frankenstein]: FAILED! => {"changed": true, "failed": true, "msg": "Docker API Error: no such id: effe8c8defc85a041802754453cd5345aa8b0781bed2cd5c6024b2e8b4aeff1c"}
```

I use it in a role `docker-registry` with the following configuration:

``` yaml

---
# roles/docker-registry/tasks/main.yml

- name: Copy config.yml for docker registry
  template: src=config.yml.j2 dest=/etc/docker_registry.yml
  notify: Restart docker registry

- name: Add docker registry container
  docker:
    name: docker-registry
    image: registry:2
    state: started
    ports: 127.0.0.1:5000:5000
    restart_policy: always
    volumes:
    - /etc/docker_registry.yml:/etc/docker/registry/config.yml
```

``` yaml

---
# roles/docker-registry/handlers/main.yml

- name: Restart docker registry
  docker:
    name: docker-registry
    image: registry:2
    state: restarted
```

The target host runs docker 1.9.1 and the ansible machine uses docker-py 1.6.0.

---

Stopping and starting the container in 2 steps works around the issue:

``` yaml

---
- name: Restart docker registry
  docker:
    name: docker-registry
    image: registry:2
    state: stopped
  notify: Start docker registry

- name: Start docker registry
  docker:
    name: docker-registry
    image: registry:2
    state: started
```

**Comment by jcassee** [comment]
I have been seeing the same problem since I started to use Ansible 2 beta. I think this happens when the container is restarted because a new image is available.

**Comment by 0x46616c6b** [comment]
I have the same problem with Ansible 2.0.0.2.

**Comment by sun-zheng-an** [comment]
I have the same problem too.

It may be caused by b42ac218df7586c7577b7f68962ce342b5fbbf21 , after reverting this commit, everything works fine.

**Comment by pleerock** [comment]
have same problem using ansible 2

**Comment by tillkuhn** [comment]
Same issue here after updating to 2.0.0.2. It even seems to happen if the image remains unchanged. Splitting the container restart into a stop and a start step works for me as well, but I hope this gets fixed soon.

**Comment by larstobi** [comment]
Same issue here after updating to 2.0.0.2.
The running container is even removed by ansible.

Here's a timeline of my scenario of a changed Logstash config file.
Ansible run prior to config file change:

```
$ ./ansible-wrapper logstash.yml opstest --pass --tags docker

PLAY ***************************************************************************

TASK [setup] *******************************************************************
ok: [logstash0]

TASK [logstash : Copy logstash config directory] *******************************
ok: [logstash0]

TASK [logstash : logstash config docker container] *****************************
ok: [logstash0]

TASK [logstash : logstash running container] ***********************************
ok: [logstash0]

PLAY RECAP *********************************************************************
logstash0                  : ok=4    changed=0    unreachable=0    failed=0
```

The logstash container is running:

```
[logstash0 ~]# docker  ps -a
CONTAINER ID        IMAGE               COMMAND                  CREATED              STATUS              PORTS                    NAMES
f89780f98204        logstash:2.1.1-1    "/docker-entrypoint.s"   About a minute ago   Up 5 seconds        0.0.0.0:5044->5044/tcp   logstash
07b718061235        busybox:1.24.1      "sh"                     8 days ago           Created                                      logstash_config
```

Now I have changed a Logstash config file. This makes the handler get notified to restart, and then it fails.

```
$ ./ansible-wrapper logstash.yml opstest --pass --tags docker

PLAY ***************************************************************************

TASK [setup] *******************************************************************
ok: [logstash0]

TASK [logstash : Copy logstash config directory] *******************************
changed: [logstash0]

TASK [logstash : logstash config docker container] *****************************
ok: [logstash0]

TASK [logstash : logstash running container] ***********************************
ok: [logstash0]

RUNNING HANDLER [logstash : Restart Logstash docker] ***************************
fatal: [logstash0]: FAILED! => {"changed": true, "failed": true, "msg": "Docker API Error: no such id: f89780f982041203e77d706b605b52384cadb599d1292b6305dd6344de14d969"}

PLAY RECAP *********************************************************************
logstash0                  : ok=4    changed=1    unreachable=0    failed=1
```

Now, the container has been completely removed, for no apparent reason(?):

```
[logstash0 ~]# docker  ps -a
CONTAINER ID        IMAGE               COMMAND             CREATED             STATUS              PORTS               NAMES
07b718061235        busybox:1.24.1      "sh"                8 days ago          Created                                 logstash_config
```

Here's my code:

```
---
# logstash/handlers/main.yml

- name: Restart Logstash docker
  docker:
    name: logstash
    image: logstash:2.1.1-1
    state: restarted
```

```
---
# logstash/tasks/main.yml

- name: Copy logstash config directory
  copy: src=conf.d dest=/config/
  notify: Restart Logstash docker
  tags:
    - docker

- name: logstash config docker container
  docker:
    name: logstash_config
    image: busybox:1.24.1
    state: present
    volumes:
    - /config:/etc/logstash
  tags:
    - docker

- name: logstash running container
  docker:
    name: logstash
    image: logstash:2.1.1-1
    command: logstash -f /etc/logstash/conf.d
    state: started
    restart_policy: always
    log_driver: syslog
    expose:
    - "5044"
    ports:
    - "5044:5044"
    volumes_from:
    - logstash_config
  tags:
    - docker
```

The docker images here have been pre-provisioned by Packer, and the VM does not have access to Docker Hub.

**Comment by cjrd** [comment]
+1 same problem for me

**Comment by satblip** [comment]
+1 same problem here

```
ansible 2.0.2.0
```

**Comment by ansibot** [comment]
@cove, @softzilla, @zfil, @ThomasSteinbach, ping. This issue is still waiting on your response.
[click here for bot help](https://github.com/jctanner/ansibullbot/blob/ISSUE_TRIAGE_CLASS/ISSUE_HELP.md)

**Comment by ozbillwang** [comment]
Need update to everyone, that stop/start doesn't work. 

If you follow up the codes which @andrenarchy pasted at beginning to stop and start container, you lost all docker container's setting, include volume, ports, entrypoint and others. 

Not sure if this is a bug in ansible docker module.

So the best command to restart docker container in ansible v2.0.x currently is to use the command directly. 

```
- name: Restart docker container
  command: docker restart "{{ docker_container_name }}"
```

**Comment by chouseknecht** [comment]
@SydOps or use the new docker_container module. The docker module is being deprecated and replaced by docker_container. You can also describe the app in Docker compose and manage it with the docker_service module.

**Comment by ansibot** [comment]
@cove, @softzilla, @zfil, @ThomasSteinbach, ping. This issue is still waiting on your response.
[click here for bot help](https://github.com/ansible/ansibullbot/blob/master/ISSUE_HELP.md)

**Comment by ozbillwang** [comment]
@chouseknecht 

Thanks for your reply, with the latest version 2.1.1.0 I have some issues which i can't use the module `docker_container` properly. 

https://github.com/ansible/ansible-modules-core/issues/4363

I have the same issue with `docker` or `docker_container` module in ansible 2.1.1.0

**Comment by chouseknecht** [comment]
#4363 was fixed for docker_container. If you drop the latest code into a library directory, you should be good.

**Comment by chouseknecht** [comment]
Actually, that might not be true. I'll double check that it's fixed.

**Comment by chouseknecht** [comment]
See my comment #4363. That issue has been fixed in the latest docker_container code.

Same thing here. The following playbook, which is almost identical to the example submitted by @larstobi, works as expected with the latest docker_container code. I am running from source. You can either run from source or place the latest docker_container module code in a library directory. 

```
- name: Test docker_container
  hosts: localhost
  gather_facts: no
  connection: local
  tasks:
    - docker_container:
        name: logstash_config
        image: busybox:1.24.1
        state: present
        volumes:
          - ./config:/etc/logstash
        debug: yes
      register: output

    - debug: var=output

    - docker_container:
        name: logstash
        image: logstash:2.1.1-1
        command: logstash -f /etc/logstash/conf.d
        restart_policy: always
        restart: yes
        expose:
          - "5044"
        ports:
          - "5044:5044"
        volumes_from:
          - logstash_config
        debug: yes
      register: output

    - debug: var=output
```

**Comment by chouseknecht** [comment]
wontfix

Existing docker.py module is being deprecated. Please try using latest docker_container.py module. To get the latest changes may require running from source or placing a copy of the most recent module code in a library directory.
